### ISSUE TITLE
JoinRelation interface weakens type in method chain

### ISSUE THREAD

**Issue by Aksigera** [Vincit/objection.js#1339]
```
class Foo extends Model {
  prop!: number;
}

async function joinRelationDoesNotSupportChaining() {
  const foo1: Foo | undefined = await Foo.query().first();
  const foo2: Foo | undefined = await Foo.query().joinRelation('any_relation').first(); // won't compile
}
```
From `JoinRelation` interface signature `foo2` inferred type would be `Model | undefined`, but, if I'm correct, it's type would never be weaker than `Foo | undefined`.

So I suggest to change typings from 
```
  interface JoinRelation {
    <QM extends Model>(relationName: string, opt?: RelationOptions): QueryBuilder<QM, QM[]>;
  }
```
to
```
  interface JoinRelation {
    <TBase extends Model, QM extends TBase>(this:QueryBuilder<TBase>, relationName: string, opt?: RelationOptions): QueryBuilder<QM, QM[]>;
  }
```

**Comment by mordof** [comment]
Ran across this. Unfortunately I don't have time to dig into it at the moment, but looking at the example change you've given @Aksigera I don't think you need `TBase`.

should work the same with:

```typescript
  interface JoinRelation {
    <QM extends Model>(this: QueryBuilder<QM>, relationName: string, opt?: RelationOptions): QueryBuilder<QM, QM[]>;
  }
```

As QM can then be derived directly from the QueryBuilder found in the `this` context.

**Comment by Aksigera** [comment]
> Ran across this. Unfortunately I don't have time to dig into it at the moment, but looking at the example change you've given @Aksigera I don't think you need `TBase`.
> 
> As QM can then be derived directly from the QueryBuilder found in the `this` context.

I don't expect `joinRelation()` to return type of `this`, that's why I added `TBase`.
In your code, output of function is of the same type as, well, input. But **join** means expanding of type, so we have such hierarchy `Model <= TBase <= QM`, where `<=` means smaller or equal.

TBase can be changed to anything, I'm not familiar with naming conventions on this project

**Comment by mordof** [comment]
adding a note here that this affects the other similar functions `rightJoinRelation`, etc. Going to tackle this issue this week

**Comment by koskimas** [comment]
Fixed in 2.0
